examples/ola-streaming-client.cpp

```diff
@@ -73,18 +79,37 @@ int main(int argc, char *argv[]) {
   StreamingClient ola_client;
   if (!ola_client.Setup()) {
     OLA_FATAL << "Setup failed";
-    exit(1);
+    exit(ola::EXIT_SOFTWARE);
   }
 
   if (FLAGS_dmx.str().empty()) {
     string input;
+    bool have_universe = false;
+    unsigned int read_universe = FLAGS_universe;
+
     while (!terminate && std::cin >> input) {
-      ola::StringTrim(&input);
-      SendDataFromString(&ola_client, FLAGS_universe, input);
+      if (!have_universe && FLAGS_universe_from_stdin) {
+        if (!ola::StringToInt(input, &read_universe, true)) {
+          OLA_FATAL << "Could not convert universe number: read " << input;
+          exit(ola::EXIT_DATAERR);
+        }
+
+        have_universe = true;
+        continue;
+      }
+
+      if (have_universe || !FLAGS_universe_from_stdin) {
+        SendDataFromString(&ola_client, read_universe, input);
```

Comment:
Does your example actually work, won't it barf trying to parse `128 2 0` as a channel value?

Reply:
In that case the interpretation is
- universe 128
  - channel 0 = 0x02
- universe 0
  - *process will block while waiting for channel values*

I'm not sure what you mean by `won't it barf trying to parse 128 2 0 as a channel value?` ... :confused:

The channel value parsing routine comes from the existing implementation. I've only added a change to parse a leading decimal value as a universe number, so existing channel data parsing behavior is still preserved.

A recording of universe 128 & 0 with a sequence of `128 2 0 1` results in:

```
OLA Show
Recording, hit Control-C to end
128 2
4915
0 1
```

Reply:
Sorry, I mean the full example from the man page is:
`1 0,255,128 2 0,255,127`

Does that actually set both universes if it's all on one line (I can't see anything to split the line).

This could just be a case of updating the example to be on two lines.

> I'm not sure what you mean by `won't it barf trying to parse 128 2 0 as a channel value?` ... confused

I mean I think it would fail to parse it as a set of channel values.

> The channel value parsing routine comes from the existing implementation. I've only added a change to parse a leading decimal value as a universe number, so existing channel data parsing behavior is still preserved.

Exactly, and that would expect chan-a,chan-b,chan-c etc. I don't see anything to stop when it hits something invalid and pass it onto the next bit of the system.

> A recording of universe 128 & 0 with a sequence of `128 2 0 1` results in:
> 
> ```
> OLA Show
> Recording, hit Control-C to end
> 128 2
> 4915
> 0 1
> ```

Why is there a bit (4s?) delay between universe 128 frame and universe 0 frame if they're on the same line?

Reply:
The 4s delay was because I wasn't entering the input on one line (separating the two universes works equally well using spaces newlines doesn't really matter, since the tokenization is done using whitespace. See (2) at https://en.cppreference.com/w/cpp/string/basic_string/operator_ltltgtgt - that's what does the "stop when it hits something invalid and pass it onto the next bit of the system." part).

The recording from the input being on one line is this:

```
1 0,255,128
0
2 0,255,127
```

Reply:
When the client reads from STDIN it would get individual strings of "1", "0,255,128", "0", "0,255,127" (see line 90 of `ola-streaming-client.cpp`).

The universe numbers are consumed, and only the comma-separated data values are passed to `SetFromString`: the SetFromString implementation doesn't need to know about the extra universe numbers.

Reply:
Ah apologies @shenghaoyang . I'd always erroneously assumed this would be just newlines and didn't realise any whitespace would work. I see the same thing does indeed work on the existing code too!